src/Host/Client/Impl/BrokerServices/WebServer.cs

```diff
@@ -11,41 +11,51 @@
 
 namespace Microsoft.R.Host.Client.BrokerServices {
     public class WebServer {
-        public string LocalHost { get; }
-        public int LocalPort { get; }
-        public string RemoteHost { get; }
-        public int RemotePort { get; }
+        private static Dictionary<int, WebServer> Servers { get; } = new Dictionary<int, WebServer>();
 
-        private readonly HttpListener _listener;
         private readonly IRemoteUriWebService _remoteUriService;
+        private HttpListener _listener;
 
-        private static Dictionary<int, WebServer> Servers { get; } = new Dictionary<int, WebServer>();
+        public string LocalHost { get; }
+        public int LocalPort { get; private set; }
+        public string RemoteHost { get; }
+        public int RemotePort { get; }
 
-        private WebServer(string remoteHostIp, int remotePort,  string baseAddress) {
+        private WebServer(string remoteHostIp, int remotePort, string baseAddress) {
             LocalHost = IPAddress.Loopback.ToString();
             RemoteHost = remoteHostIp;
             RemotePort = remotePort;
 
             _remoteUriService = new RemoteUriWebService(baseAddress);
-            Random r = new Random();
-
-            // if remote port is between 10000 and 32000, select a port in the same range.
-            // R Help uses ports in that range.
-            int localPortMin = (RemotePort >= 10000 && RemotePort <= 32000)? 10000: 49152;
-            int localPortMax = (RemotePort >= 10000 && RemotePort <= 32000) ? 32000 : 65535;
-
-            while(true) {
-                _listener = new HttpListener();
-                LocalPort = r.Next(localPortMin, localPortMax);
-                _listener.Prefixes.Add(Invariant($"http://{LocalHost}:{LocalPort}/"));
-                try {
-                    _listener.Start();
-                } catch (HttpListenerException) {
-                    _listener.Close();
-                    continue;
+        }
+
+        public Task InitializeAsync(CancellationToken ct) {
+            return Task.Run(() => {
```

Comment:
#2500. Original code did the loop in sync without option to cancel. From the dump it appears loop was stuck in HttpListener.AddPrefix.